spiffe/tls_peer.go

```diff
@@ -0,0 +1,319 @@
+	return func(p *TLSPeer) error {
+		p.log = log
+		return nil
+	}
+}
+
+type TLSPeer struct {
+	log    Logger
+	addr   string
+	client *workload.X509SVIDClient
+
+	readyOnce sync.Once
+	ready     chan struct{}
+
+	mu    sync.RWMutex
+	cert  *tls.Certificate
+	roots map[string]*x509.CertPool
+}
+
+func NewTLSPeer(opts ...TLSPeerOption) (*TLSPeer, error) {
+	p := &TLSPeer{
+		ready: make(chan struct{}),
+	}
+
+	for _, opt := range opts {
+		if err := opt(p); err != nil {
+			return nil, err
+		}
+	}
+
+	if p.log == nil {
+		p.log = nullLogger{}
+	}
+
+	var dialOpts []workload.DialOption
+	if p.addr != "" {
+		dialOpts = append(dialOpts, workload.WithAddr(p.addr))
+	}
+
+	client, err := workload.NewX509SVIDClient(&tlsPeerWatcher{p: p}, dialOpts...)
+	if err != nil {
+		return nil, err
+	}
+	client.Start()
+
+	p.client = client
+	return p, nil
+}
+
+func (p *TLSPeer) Close() error {
+	return p.client.Stop()
+}
+
+func (p *TLSPeer) WaitUntilReady(ctx context.Context) error {
+	select {
+	case <-p.ready:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (p *TLSPeer) GetCertificate() (*tls.Certificate, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	if p.cert == nil {
+		return nil, errors.New("workload does not have a certificate yet")
+	}
+	return p.cert, nil
+}
+
+func (p *TLSPeer) GetRoots() (map[string]*x509.CertPool, error) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	if p.roots == nil {
+		return nil, errors.New("workload does not have roots yet")
+	}
+	return p.roots, nil
+}
+
+func (p *TLSPeer) updateX509SVIDs(svids *workload.X509SVIDs) {
+	p.log.Debugf("X509SVID workload API update received")
+
+	// Use the default SVID for now
+	// TODO: expand SVID selection options
+	svid := svids.Default()
+	_, trustDomainID, err := internal.GetIDsFromCertificate(svid.Certificates[0])
+	if err != nil {
+		p.onError(errors.New("unable to parse IDs from X509-SVID update"))
+		return
+	}
+
+	cert := &tls.Certificate{
+		Certificate: make([][]byte, 0, len(svid.Certificates)),
+		PrivateKey:  svid.PrivateKey,
+	}
+	for _, svidCert := range svid.Certificates {
+		cert.Certificate = append(cert.Certificate, svidCert.Raw)
+	}
+
+	roots := make(map[string]*x509.CertPool)
+	for federatedDomainID, federatedDomainPool := range svid.FederatedTrustBundlePools {
+		roots[federatedDomainID] = federatedDomainPool
+	}
+	roots[trustDomainID] = svid.TrustBundlePool
+
+	p.mu.Lock()
+	p.cert = cert
+	p.roots = roots
+	p.mu.Unlock()
+
+	p.readyOnce.Do(func() {
+		close(p.ready)
+	})
+}
+
+func (p *TLSPeer) onError(err error) {
+	p.log.Errorf("%v", err)
+}
+
+type tlsPeerWatcher struct {
+	p *TLSPeer
+}
+
+func (w *tlsPeerWatcher) UpdateX509SVIDs(svids *workload.X509SVIDs) {
+	w.p.updateX509SVIDs(svids)
+}
+
+func (w *tlsPeerWatcher) OnError(err error) {
+	w.p.onError(err)
+}
+
+func (p *TLSPeer) Dial(ctx context.Context, network, address string, expectPeer ExpectPeerFunc) (net.Conn, error) {
+	conn, err := new(net.Dialer).DialContext(ctx, network, address)
```

Comment:
I didn't want to artificially restrict the transport protocol that workloads use talk to each other over. If it's good enough for "net/tls", it seemed good enough to me.

Reply:
Hmm ok. I am not sure what checks or handling tls.Dial() does on the network string. I don't see any in the go source, but that doesn't mean there's not any. Note that the difference with that is here we are passing network string to net.Dialer and not to tls.Dial where the TLS package may have some special considerations.

For example, from the docs, it seems that net.Dialer will happily open an ICMP "connection" if it has the privileges to do so. But there is no mention of ICMP being supported in the TLS package. I guess my concern here is I wouldn't want to leave a path for this to be misused that ends up leaving someone vulnerable.

The crypto/tls package has Dial equivalents as you pointed out - is it possible to use those directly instead of using net.Dial?

Reply:
Talked offline and determined that using the currently dialing methods was ok for now.